### PR TITLE
feat: preferences controller to base controller v2

### LINKED ISSUE #25917: Convert PreferencesController to TypeScript, and migrate to BaseController v2
### What is this about?

Following the [Wallet Framework team's OKRs for Q3 2024](https://docs.google.com/document/d/1JLEzfUxHlT8lw8ntgMWG0vQb5BAATcrYZDj0wRB2ogI/edit#heading=h.kzzai3cfecro), we want to bring `AlertController` up to date with our latest controller patterns.

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

- `PreferencesController` inherits from `BaseController` (out of `@metamask/base-controller`).
  - The `PreferencesController` constructor signature takes a partial `state` option instead of `initState`.
  - The default state is extracted to a `getDefaultPreferencesControllerState` function.
  - The `PreferencesController` constructor is updated to `super` to the superclass.
  - `PreferencesController` uses `this.messagingSystem` to refer to the messenger.
  - `PreferencesController` no longer registers `PreferencesController:getState` or an initial event payload for `PreferencesController:stateChange` (as this is done automatically).
  - `PreferencesController` no longer takes or sets a `network` property.
  - `PreferencesController` uses `NetworkController:getState` to get `networkConfigurationsByChainId` instead of taking a `networkConfigurationsByChainId` option.
  - `this.state` is used to access state instead of `this.store.getState`.
  - `this.update` is used to update state instead of `this.store.updateState`.
- `PreferencesController` is converted to TypeScript.
  - The `PreferencesControllerMessenger` type exists and expects `AccountsController:stateChange`, `NetworkController:getState`, `AccountsController:getAccountByAddress`, `AccountsController:setSelectedAccount`, and `AccountsController:setAccountName` to be allowed.
  - `PreferencesControllerState` exists and represents the current shape of the state object.
  - Types for unconverted files are filled in.
- The tests are updated to follow suit.
  - `preferencesController.state` is used to access state instead of `preferencesController.store.getState()`.
  - ❌ The use of `beforeEach` is _not_ corrected. We can solve this in another PR.

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

- https://github.com/MetaMask/core/issues/1509
- https://github.com/MetaMask/metamask-extension/issues/23017

### PR DESCRIPTION
**Pull request by kanthesha**
## **Description**
In this PR, we want to bring PreferencesController up to date with our latest controller patterns by upgrading to BaseControllerV2.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27398?quickstart=1)

## **Related issues**

Fixes: #25917 

## **Manual testing steps**
Use case 1
1. Install the previous release
2. Complete user onboarding
3. Go to settings and change couple of user settings. For example language, currency and theme.
4. Close and disable MM in the extension
5. Checkout the version with these changes
6. Build and login
7. Make sure, the user preferences set earlier are still there

Use case 2
1. Disable all the MM extensions
2. Install the version with these changes
3. When you click on MM, the default language should be English
4. Complete user onboarding
5. Go to settings and change couple of user settings. For example language, currency and theme.
6. Close and disable and enable the MM in extension. which forces user to login MM in the extension
7. Once you login again, make sure, the user preferences set earlier are still there

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/controllers/app-state.js [hunk 1]
```diff
@@ -100,7 +106,8 @@ export default class AppStateController extends EventEmitter {
         }),
     );
 
-    const { preferences } = preferencesStore.getState();
+    const { preferences } = preferencesController.state;
```
Comment (mikesposito) [inline app/scripts/controllers/app-state.js:109]: Can we use the messenger here, so we avoid passing the controller instance to constructor options?

File: app/scripts/controllers/preferences-controller.ts
Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:1]: nit: The current naming convention for controllers in the extension is without a `-controller` suffix, with few (if not one) exceptions like `mmi-controller` and the main `metamask-controller`. Thoughts on keeping the old name (`preferences.ts`)?

File: app/scripts/lib/account-tracker.js
Comment (mikesposito) [inline app/scripts/lib/account-tracker.js:1]: It looks like this controller is using a messenger even though it does not implement BaseController. Do you think we can take this chance to remove the `preferencesController` construction option and use `controllerMessenger` instead?

File: development/ts-migration-dashboard/files-to-convert.json [hunk 2]
```diff
@@ -40,8 +40,6 @@
   "app/scripts/controllers/permissions/selectors.test.js",
   "app/scripts/controllers/permissions/specifications.js",
   "app/scripts/controllers/permissions/specifications.test.js",
-  "app/scripts/controllers/preferences.js",
-  "app/scripts/controllers/preferences.test.js",
```
Comment (mikesposito) [inline development/ts-migration-dashboard/files-to-convert.json:44]: Wow, great catch! I see that this file is very outdated (there's even network controller which is long gone from extension)

File: app/scripts/controllers/preferences-controller.ts [hunk 3]
```diff
@@ -132,14 +136,15 @@ export type PreferencesControllerState = {
...
+  lostIdentities: Record<string, AccountIdentityEntry>;
   forgottenPassword: boolean;
   preferences: Preferences;
   ipfsGateway: string;
   isIpfsGatewayEnabled: boolean;
   useAddressBarEnsResolution: boolean;
   ledgerTransportType: LedgerTransportTypes;
-  snapRegistryList: Record<string, object>;
+  // TODO: replace never with proper type
+  snapRegistryList: Record<string, never>;
```
Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:147]: curious: why was `never` used here?

File: app/scripts/controllers/preferences-controller.ts [hunk 4]
```diff
@@ -149,125 +154,293 @@ export type PreferencesControllerState = {
...
+  snapsAddSnapAccountModalDismissed: false,
+  ///: END:ONLY_INCLUDE_IF
+  useExternalNameSources: true,
+  useTransactionSimulations: true,
+  enableMV3TimestampSave: true,
+  // Turning OFF basic functionality toggle means turning OFF this useExternalServices flag.
+  // Whenever useExternalServices is false, certain features will be disabled.
+  // The flag is true by Default, meaning the toggle is ON by default.
+  useExternalServices: true,
+});
```
Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:236]: Can we add the return type here so that we make sure to always return a valid state?

```suggestion
export const getDefaultPreferencesControllerState =
  (): PreferencesControllerState => ({
    selectedAddress: '',
    useBlockie: false,
    useNonceField: false,
    usePhishDetect: true,
    dismissSeedBackUpReminder: false,
    useMultiAccountBalanceChecker: true,
    useSafeChainsListValidation: true,
    // set to true means the dynamic list from the API is being used
    // set to false will be using the static list from contract-metadata
    useTokenDetection: true,
    useNftDetection: true,
    use4ByteResolution: true,
    useCurrencyRateCheck: true,
    useRequestQueue: true,
    openSeaEnabled: true,
    securityAlertsEnabled: true,
    watchEthereumAccountEnabled: false,
    bitcoinSupportEnabled: false,
    bitcoinTestnetSupportEnabled: false,
    ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
    addSnapAccountEnabled: false,
    ///: END:ONLY_INCLUDE_IF
    advancedGasFee: {},

    // WARNING: Do not use feature flags for security-sensitive things.
    // Feature flag toggling is available in the global namespace
    // for convenient testing of pre-release features, and should never
    // perform sensitive operations.
    featureFlags: {},
    incomingTransactionsPreferences: {
      ...mainNetworks,
      ...testNetworks,
    },
    knownMethodData: {},
    currentLocale: '',
    identities: {},
    lostIdentities: {},
    forgottenPassword: false,
    preferences: {
      autoLockTimeLimit: undefined,
      showExtensionInFullSizeView: false,
      showFiatInTestnets: false,
      showTestNetworks: false,
      smartTransactionsOptInStatus: null, // null means we will show the Smart Transactions opt-in modal to a user if they are eligible
      useNativeCurrencyAsPrimaryCurrency: true,
      hideZeroBalanceTokens: false,
      petnamesEnabled: true,
      redesignedConfirmationsEnabled: true,
      redesignedTransactionsEnabled: true,
      featureNotificationsEnabled: false,
      isRedesignedConfirmationsDeveloperEnabled: false,
      showConfirmationAdvancedDetails: false,
      showMultiRpcModal: false,
    },
    // ENS decentralized website resolution
    ipfsGateway: IPFS_DEFAULT_GATEWAY_URL,
    isIpfsGatewayEnabled: true,
    useAddressBarEnsResolution: true,
    // Ledger transport type is deprecated. We currently only support webhid
    // on chrome, and u2f on firefox.
    ledgerTransportType: window.navigator.hid
      ? LedgerTransportTypes.webhid
      : LedgerTransportTypes.u2f,
    snapRegistryList: {},
    theme: ThemeType.os,
    ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
    snapsAddSnapAccountModalDismissed: false,
    ///: END:ONLY_INCLUDE_IF
    useExternalNameSources: true,
    useTransactionSimulations: true,
    enableMV3TimestampSave: true,
    // Turning OFF basic functionality toggle means turning OFF this useExternalServices flag.
    // Whenever useExternalServices is false, certain features will be disabled.
    // The flag is true by Default, meaning the toggle is ON by default.
    useExternalServices: true,
  });
```

File: app/scripts/controllers/preferences-controller.ts [hunk 5]
```diff
@@ -149,125 +154,293 @@ export type PreferencesControllerState = {
...
 
-  private messagingSystem: PreferencesControllerMessenger;
+/**
+ * {@link PreferencesController}'s metadata.
+ *
+ * This allows us to choose if fields of the state should be persisted or not
+ * using the `persist` flag; and if they can be sent to Sentry or not, using
+ * the `anonymous` flag.
+ */
+const controllerMetadata = {
```
Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:245]: I see that none of these flags is persisted. Should all these user preferences be volatile and non-anonymous?

File: app/scripts/controllers/preferences-controller.ts [hunk 6]
```diff
@@ -132,14 +136,16 @@ export type PreferencesControllerState = {
...
   forgottenPassword: boolean;
   preferences: Preferences;
   ipfsGateway: string;
   isIpfsGatewayEnabled: boolean;
   useAddressBarEnsResolution: boolean;
   ledgerTransportType: LedgerTransportTypes;
-  snapRegistryList: Record<string, object>;
+  // TODO: Replace `any` with type
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  snapRegistryList: Record<string, any>;
```
Comment (MajorLift) [inline app/scripts/controllers/preferences-controller.ts:148]: Is an `any` necessary here? Looking at the mock object for `snapRegistryList` https://github.com/MetaMask/metamask-extension/blob/79feca9a0f422c924af314b01a72d4251b22d087/test/data/mock-state.json#L607, it looks like `Json` would be wide enough.

```suggestion
  // TODO: Replace with correct type
  snapRegistryList: Record<string, Json>;
```

File: app/scripts/controllers/preferences-controller.ts [hunk 7]
```diff
@@ -149,125 +155,294 @@ export type PreferencesControllerState = {
...
+    bitcoinTestnetSupportEnabled: false,
+    ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
+    addSnapAccountEnabled: false,
+    ///: END:ONLY_INCLUDE_IF
+    advancedGasFee: {},
+
+    // WARNING: Do not use feature flags for security-sensitive things.
+    // Feature flag toggling is available in the global namespace
+    // for convenient testing of pre-release features, and should never
+    // perform sensitive operations.
```
Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:190]: do you think it would make sense to move these warnings and comments to the `PreferencesControllerState` type declaration?

File: app/scripts/controllers/preferences-controller.ts (hunk 5 above)
Comment (mikesposito) [reply app/scripts/controllers/preferences-controller.ts:247]: On `develop`, PreferencesController.store is passed to `this.store` in    Metamask Controller: https://github.com/MetaMask/metamask-extension/blob/eea557d94674fc8c20b04a3966f97f6211990aa9/app/scripts/metamask-controller.js#L2345

Which means that currently they are all persisted

File: app/scripts/controllers/preferences-controller.ts [hunk 8]
```diff
@@ -149,125 +159,289 @@ export type PreferencesControllerState = {
...
+    // Ledger transport type is deprecated. We currently only support webhid
+    // on chrome, and u2f on firefox.
+    ledgerTransportType: window.navigator.hid
+      ? LedgerTransportTypes.webhid
+      : LedgerTransportTypes.u2f,
+    snapRegistryList: {},
+    theme: ThemeType.os,
+    ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
+    snapsAddSnapAccountModalDismissed: false,
+    ///: END:ONLY_INCLUDE_IF
```
Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:229]: This `///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)` represents a feature flag, so `snapsAddSnapAccountModalDismissed` may not be present in this default state.

Perhaps we should mark `snapsAddSnapAccountModalDismissed` as optional in the `PreferencesControllerState` type?

File: app/scripts/controllers/preferences-controller.ts [hunk 9]
```diff
@@ -149,125 +159,289 @@ export type PreferencesControllerState = {
...
+    useCurrencyRateCheck: true,
+    useRequestQueue: true,
+    openSeaEnabled: true,
+    securityAlertsEnabled: true,
+    watchEthereumAccountEnabled: false,
+    bitcoinSupportEnabled: false,
+    bitcoinTestnetSupportEnabled: false,
+    ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
+    addSnapAccountEnabled: false,
+    ///: END:ONLY_INCLUDE_IF
```
Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:188]: `addSnapAccountEnable` is under a feature flag, perhaps we should mark also this as optional in the state? (See comment related to `snapsAddSnapAccountModalDismissed`)

File: app/scripts/controllers/preferences-controller.ts [hunk 10]
```diff
@@ -149,125 +159,289 @@ export type PreferencesControllerState = {
...
+      metadata: controllerMetadata,
+      name: controllerName,
+      state: {
+        ...getDefaultPreferencesControllerState(),
+        incomingTransactionsPreferences: {
+          ...mainNetworks,
+          ...addedNonMainNetwork,
+          ...testNetworks,
+        },
+        currentLocale: initLangCode ?? '',
```
Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:442]: Do you think we can move these to `getDefaultPreferencesControllerState`?

Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:442]: Shouldn't this be part of the `state` passed from MetamaskController? I see that `currentLocale` is also part of the persisted state

Comment (mikesposito) [reply app/scripts/controllers/preferences-controller.ts:442]: This to say: do you think it would make sense to simply remoive `initLangCode` from the constructor options?

File: app/scripts/controllers/preferences-controller.ts [hunk 11]
```diff
@@ -149,125 +159,289 @@ export type PreferencesControllerState = {
...
+   * Constructs a Preferences controller.
+   *
+   * @param options - the controller options
+   * @param options.networkConfigurationsByChainId - The network configurations
+   * @param options.initLangCode - The language code
+   * @param options.messenger - The controller messenger
+   * @param options.state - The initial controller state
+   */
+  constructor({
+    networkConfigurationsByChainId,
```
Comment (mikesposito) [inline app/scripts/controllers/preferences-controller.ts:420]: Do you think we can get this with a messenger call, so we can remove this option?

File: app/scripts/controllers/app-state.js (hunk 1 above)
Comment (cryptodev-2s) [reply app/scripts/controllers/app-state.js:109]: @kanthesha we don't need a new ticket for this it should be done part of the work on this ticket can you instead add a comment or update the description https://github.com/MetaMask/metamask-extension/issues/25916

File: app/scripts/metamask-controller.js [hunk 12]
```diff
@@ -7031,30 +7050,6 @@ export default class MetamaskController extends EventEmitter {
...
-    if (currentState.incomingTransactionsPreferences?.[chainId]) {
-      this.txController.startIncomingTransactionPolling();
-    } else {
-      this.txController.stopIncomingTransactionPolling();
-    }
-
-    this.#checkTokenListPolling(currentState, previousState);
-
-    // TODO: Remove once the preferences controller has been replaced with the core monorepo implementation
-    this.controllerMessenger.publish(
```
Comment (cryptodev-2s) [inline app/scripts/metamask-controller.js:7051]: Why did you remove this this.controllerMessenger.publish(
      'PreferencesController:stateChange',
      currentState,
      [],
    ); ?

Comment (cryptodev-2s) [reply app/scripts/metamask-controller.js:7051]: I'm unsure if it's still required, but it was introduced in this PR (https://github.com/MetaMask/metamask-extension/pull/23919/files) to ensure we're updating the core's implementation of the PreferencesController, which is a peer dependency of the AssetsController. We should ensure this doesn't block that flow.

Comment (mikesposito) [reply app/scripts/metamask-controller.js:7051]: > as both controllers—using the same name—would register the same event listeners, potentially causing conflicts.

I believe that the idea here is that a client would never instantiate both of the `PreferencesController` (extension implementation and core implementation), but that at any point in time there's a controller registered with the name `PreferencesController`. 

If we rename PreferencesController to something else, other controllers would fail to receive events from the controller used by the extension.

```typescript
// TODO: Remove once the preferences controller has been replaced with the core monorepo implementation
```
My suspect is that this comment is misleading: it should say that the manual publishing should be removed once `PreferencesController` (core or extension, doesn't matter) starts using the same core's messenger.

Comment (mikesposito) [reply app/scripts/metamask-controller.js:7051]: Before this PR, `PreferencesController` was still using `ObservableStore`, therefore it was not autonomously publishing state change events through the messenger

Comment (cryptodev-2s) [reply app/scripts/metamask-controller.js:7051]: Before this PR, we were publishing autonomously to the core's implementation, as the call was triggered by subscribing to changes in the observable store from the extension's PreferencesController.
check https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/metamask-controller.js#L2833
and https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/metamask-controller.js#L7053

Comment (cryptodev-2s) [reply app/scripts/metamask-controller.js:7051]: The comment is correct, as it states that this should only be removed once we eliminate the Preferences extension implementation and fully replace it with the core's implementation.

Comment (cryptodev-2s) [reply app/scripts/metamask-controller.js:7051]: "If we rename PreferencesController to something else, other controllers would fail to receive events from the controller used by the extension." -> Yes, but currently, we don't have a messenger for the extension PreferencesController. If we choose a different name, all controllers that need to interact with the extension's preferences will simply use the new name we assign to it.

Comment (cryptodev-2s) [reply app/scripts/metamask-controller.js:7055]: If we assume what you're saying is correct, with your change, when `PreferencesController:getState` is called from any controller or another flow, which state will it return?

Comment (Gudahtt) [reply app/scripts/metamask-controller.js:7055]: Interesting. So the options we have are:

## Give the two controllers the same name

Pros:
* It's more similar to how things work today
Cons:
* Less type-safe, as the differences between the extension and core Preferences controllers are obscured. Other core controllers may expect properties to exist that do not exist, leading to errors.

## Give the two controllers different names, and preserve this publish event as an "adapter" for core controllers that expect this event

Pros:
* More type safe
Cons:
* Higher change of introducing unintentional functional changes

Comment (Gudahtt) [reply app/scripts/metamask-controller.js:7055]: I was assuming we'd proceed with the first option, which is more similar to how things already work. But it's not a great situation to be in. The mismatch in types is not ideal.

Comment (cryptodev-2s) [reply app/scripts/metamask-controller.js:7055]: Thanks, @Gudahtt ! That means the publish event was ineffective and essentially unnecessary from the start.

Comment (mikesposito) [reply app/scripts/metamask-controller.js:7075]: As core PreferencesController does not publish other events rather than `stateChange`, could this be a solution in between the two proposed by @Gudahtt ?

```typescript
import type { PreferencesState } from '@metamask/preferences-controller';

export type PreferencesControllerState = PreferencesState & {
  // .. extend core PreferencesState with extension-specific keys
};
```

This would ensure enough overlap between the two state versions to keep other controllers functioning

**Review by cryptodev-2s — DISMISSED** [review]
LGTM; Great work!

File: app/scripts/controllers/preferences-controller.ts (hunk 3 above)
Comment (mikesposito) [reply app/scripts/controllers/preferences-controller.ts]: I see this was changed to `Json`, nice!

File: app/scripts/metamask-controller.js [hunk 13]
```diff
@@ -1684,7 +1694,18 @@ export default class MetamaskController extends EventEmitter {
...
+      controllerMessenger: this.controllerMessenger.getRestricted({
+        name: 'AccountTracker',
+        allowedActions: ['AccountsController:getSelectedAccount'],
+        allowedEvents: [
+          'AccountsController:selectedEvmAccountChange',
+          'OnboardingController:stateChange',
+          'KeyringController:accountRemoved',
+        ],
+      }),
+      initState: { accounts: {} },
```
Comment (cryptodev-2s) [inline app/scripts/metamask-controller.js:1708]:
```suggestion
```

File: app/scripts/lib/account-tracker.js
Comment (cryptodev-2s) [reply app/scripts/lib/account-tracker.js:1]: @mikesposito changes are done here https://github.com/MetaMask/metamask-extension/pull/27711

**Review by cryptodev-2s — DISMISSED** [review]

**Review by mikesposito — DISMISSED** [review]
Amazing work! I tested the changes migrating from `develop` and installing directly from this branch - no issue observed

LGTM!

File: app/scripts/controllers/preferences-controller.ts [hunk 14]
```diff
@@ -114,176 +123,356 @@ export type Preferences = {
...
+    useExternalNameSources: true,
+    useTransactionSimulations: true,
+    enableMV3TimestampSave: true,
+    // Turning OFF basic functionality toggle means turning OFF this useExternalServices flag.
+    // Whenever useExternalServices is false, certain features will be disabled.
+    // The flag is true by Default, meaning the toggle is ON by default.
+    useExternalServices: true,
+    // from core PreferencesController
+    isMultiAccountBalancesEnabled: true,
+    showIncomingTransactions: {
```
Comment (DDDDDanica) [inline app/scripts/controllers/preferences-controller.ts:248]: This seems to overlap with 
```
    incomingTransactionsPreferences: {
      ...mainNetworks,
      ...testNetworks,
    },
```
Can we combine both?

Comment (DDDDDanica) [reply app/scripts/controllers/preferences-controller.ts:248]: Good to know it is taken care of. Do ping me if you need some 👁️ for that ticket to refine or review ~

**Comment by DDDDDanica** [comment]
LGTM !

**Review by DDDDDanica — DISMISSED** [review]

**Review by cryptodev-2s — DISMISSED** [review]

**Review by mikesposito — DISMISSED** [review]

**Review by DDDDDanica — DISMISSED** [review]

**Review by DDDDDanica — APPROVED** [review]

**Review by cryptodev-2s — APPROVED** [review]
